rush.md

```diff
@@ -0,0 +1,374 @@
+
+ **Note:** *The description of `OrderData` provided below is nowhere near to what happens in Rush's code. This description is optimized for ease of understanding the main concept. An efficient implementation of the idea is another pair of shoes. In fact, the name `OrderData` is not even used in the code -- the ordered data instead lands in an output stream that is produced asynchronously.*
+
+
+#### 2.3.1 Description of OrderData.
+
+For a unit `U` in a certain dag let us denote by `below(U)` the set of all units that can be reached from `U` by following 0 or more forward edges from `U`, so this set consists of `U`, parents of `U`, parents of parents of `U`, and so on. Below we describe `OrderData(D)` based on a procedure `Head(r, D)` that given a round `r` and dag `D` outputs an element of type `Option<Unit>` thus either `None` or `Some(U)` where `U` is some unit in `D` of round `r`. The description of  `Head(r, D)` follows later.
+
+```
+def OrderData(D):
+	order = []
+	batch = []
+	for r = 0, 1, 2, ... :
+		if Head(r, D) = Some(U):
+			new_batch = below(U)
+			remove from new_batch the contents of batch[0], batch[1], ..., batch[r-1]
+			sort new_batch using any canonical ordering
+			batch[r] = new_batch
+			let [U_1, U_2, ..., U_k] = new_batch
+			order = order ++ [U_1.data, U_2.data, ..., U_k.data]
+		else:
+			break
+	return order
+```
+ The `OrderData` algorithm is indeed quite simple: we choose a `Head` unit from each consecutive round (until it is no more possible) and generate a new batch of units, by taking all units below the head, and removing units from previous batches. The ordering is simply the data items from the units in subsequent batches, where each batch is ordered arbitrarily (yet using some deterministic ordering).  It is not hard to observe that `OrderData` has the desired monotonicity property as long as the black-box `Head(r, D)` is also "monotone", in the sense that:
+
+  **If `D` is a subset of `D'` and `Head(r, D)=Some(U)` then  `Head(r, D')=Some(U)`.**
+
+  We proceed to describing `Head(r, D)`
+
+#### 2.3.2 Description of Head.
+
+We will again describe `Head` as a procedure using another black-box: `Decide(U, D)` which takes a unit `U` and a dag `D` and outputs a value of type `Option<bool>`, so either a boolean "decision" `Some(b)` or `None`. We assume here that `Decide(U, D)` satisfies an analogous property to the one mentioned for `Head` above.
+
+```
+def Head(r, D):
+	let [U_1, U_2, ..., U_s] be the units in D of round r sorted according to some canonical order
+	// The order is allowed to depend on r and on Head(r-1, D)
+	for l = 1, 2, ..., s:
+		if Decide(U_l, D) == Some(b):
+			if b==true:
+				return U_l
+			else:
+				continue the loop
+		else:
+			return None
+```
```

Comment:
What if all `Decide(U_l, D)` return false

Reply:
Actually "continue" is when Decide() returns Some(false)

Reply:
In the current version if all `Decide(U_l, D)` return `Some(false)`, we reach the end of the function and there is no `return` there, and we should return `Option<Unit>`.

Also, in the `if b==true` case, we should wrap `U_l` with `Some`

Reply:
Ok, there were indeed some mistakes there. I added a fix: it is required that the Dag is high enough for Head to execute (i.e., at least round r+3 reached) then, the code cannot exit the for loop without returning `None`. Added also a comment at the end of the function.